common/lib/capa/capa/capa_problem.py

```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+                element_to_be_deleted = responsetype_label_tag
+            elif 'label' in inputfields[0].attrib:
+                # Extract label value from label attribute
+                # This is the case when we have a problem
+                # * with multiple questions without separation
+                # * single question with old XML format only
+
+                label = inputfields[0].attrib['label']
+                # Get first <p> tag before responsetype
+                p_tag = response.xpath('preceding-sibling::p[1]')
+
+                if p_tag:
+                    # It may be possible that label attribute value doesn't match with <p> tag
+                    # This happens when author updated the question <p> tag directly in XML but
+                    # didn't changed the label attribute value. In this case we will consider the
+                    # first <p> tag before responsetype as question.
+                    if label != p_tag[0].text:
+                        label = p_tag[0].text
+                    element_to_be_deleted = p_tag[0]
+            else:
+                # neither <label> tag nor label attribute is present inside responsetype
+                # This is the case when we have a problem with multi-questions without --- having markdown
+                # find the immediate <label> tag before the responsetype. also delete it avoid rendering twice
+                label_tag = response.xpath("preceding-sibling::label[1]")
+                if label_tag:
+                    label = label_tag[0].text
+                    element_to_be_deleted = label_tag[0]
+
+            label = label.strip() or DEFAULT_QUESTION_TEXT
+
+            # delete label or p element only if responsetype is fully accessible
+            if response.tag in ACCESSIBLE_CAPA_RESPONSE_TYPES and element_to_be_deleted is not None:
+                element_to_be_deleted.getparent().remove(element_to_be_deleted)
+
+            # for non-accessible responsetypes it may be possible that label attribute is not present
+            # in this case pass an empty label. remember label attribute is only used as value for aria-label
+            if response.tag not in ACCESSIBLE_CAPA_RESPONSE_TYPES and label == DEFAULT_QUESTION_TEXT:
+                label = ''
+
+            # Extract descriptions and set unique id on each description tag
+            description_tags = response.findall('description')
+            description_id = 1
+            descriptions = OrderedDict()
+            for description in description_tags:
+                descriptions[
+                    "description_%s_%i_%i" % (self.problem_id, response_id, description_id)
+                ] = description.text
+                response.remove(description)
+                description_id += 1
+
+            problem_data[question_id] = {
+                'label': label,
+                'description_ids': ' '.join(descriptions.keys()),
```

Comment:
Can we use keys of `descriptions` wherever we're using `description_ids`, this way we don't need to maintain this separate description_ids.